include/xtensor/xsort.hpp

```diff
@@ -772,6 +780,335 @@ namespace xt
+            // Both interpolation coefficients, 1-gamma and gamma
+            auto const omg_g = concatenate(xtuple(T(1) - (p_n_m - j), p_n_m - j));
+            return std::make_pair(eval(k_kp1), eval(omg_g));
+        }
+
+        // TODO should implement unsqueeze rather
+        template <class S>
+        inline auto unsqueeze_shape(S const& shape, std::size_t axis)
+        {
+            XTENSOR_ASSERT(axis <= shape.size());
+            auto new_shape = xtl::forward_sequence<xt::svector<std::size_t>, decltype(shape)>(shape);
+            new_shape.insert(new_shape.begin() + axis, 1);
+            return new_shape;
+        }
+    }
+
+    /**
+     * Compute quantiles over the given axis.
+     *
+     * In a sorted array represneting a distribution of numbers, the quantile of a probability ``p``
+     * is the the cut value ``q`` such that a fraction ``p`` of the distribution is lesser or equal
+     * to ``q``.
+     * When the cutpoint falls between two elemnts of the sample distribution, a interpolation is
+     * computed using the @p alpha and @p beta coefficients, as descripted in
+     * (Hyndman and Fan, 1996).
+     *
+     * The algorithm partially sorts entries in a copy along the @p axis axis.
+     *
+     * @ingroup xt_xsort
+     * @param e Expression containing the distribution over which the quantiles are computed.
+     * @param probas An list of probability associated with each desired quantiles.
+     *        All elements must be in the range ``[0, 1]``.
+     * @param axis The dimension in which to compute the quantiles, *i.e* the axis representing the
+     *        distribution.
+     * @param alpha Interpolation parameter. Must be in the range ``[0, 1]]``.
+     * @param beta Interpolation parameter. Must be in the range ``[0, 1]]``.
+     * @tparam T The type in which the quatile are computed.
+     * @return An expression with as many dimensions as the input @p e.
+     *         The first axis correspond to the quantiles.
+     *         The other axes are the axes that remain after the reduction of @p e.
+     * @see (Hyndman and Fan, 1996) R. J. Hyndman and Y. Fan,
+     *      "Sample quantiles in statistical packages", The American Statistician,
+     *      50(4), pp. 361-365, 1996
+     * @see https://en.wikipedia.org/wiki/Quantile
+     */
+    template <class T = double, class E, class P>
+    inline auto quantile(E&& e, P const& probas, std::ptrdiff_t axis, T alpha, T beta)
+    {
+        // Check inputs
+        if (any((probas < 0.) || (1. < probas)))
+        {
+            XTENSOR_THROW(std::runtime_error, "p must be between 0 and 1");
+        }
+        if ((alpha < 0) || (1 < alpha))
+        {
+            XTENSOR_THROW(std::runtime_error, "alpha must be between 0 and 1");
+        }
+        if ((beta < 0) || (1 < beta))
+        {
+            XTENSOR_THROW(std::runtime_error, "beta must be between 0 and 1");
+        }
```

Comment:
I think these checks shoud be disabled by default, kind of what we do for checking index bounds in the access operator.

Reply:
The policy for handling checks/errors is a bit obscure to me: `XTENSOR_THROW`, `XTENSOR_ASSERT`, `xt::check_policy`.

I changed it to `XTENSOR_ASSERT`, is that better?

Reply:
Yep, that's perfect! XTENSOR_ASSERT perfroms checks only if you enable it, `XTENSOR_THROW` allows to enable/disable exceptoins at build time.